com.unity.ml-agents/Runtime/MLAgentsManager.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+using System.Linq;
+using UnityEngine;
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+
+namespace Unity.MLAgents
+{
+#if UNITY_EDITOR
+    [InitializeOnLoad]
+#endif
+    internal static class MLAgentsManager
+    {
+        internal static event Action OnSettingsChange;
+        internal const string EditorBuildSettingsConfigKey = "com.unity.ml-agents.settings";
+        private static MLAgentsSettings s_Settings;
+
+
+        public static MLAgentsSettings Settings
+        {
+            get
+            {
+                if (s_Settings == null)
+                {
+                    Initialize();
+                }
+                return s_Settings;
+            }
+            set
+            {
+                Debug.Assert(value != null);
+#if UNITY_EDITOR
+                if (!string.IsNullOrEmpty(AssetDatabase.GetAssetPath(value)))
+                {
+                    EditorBuildSettings.AddConfigObject(EditorBuildSettingsConfigKey, value, true);
+                }
+#endif
+                s_Settings = value;
+                ApplySettings();
+            }
+        }
+
+        static MLAgentsManager()
+        {
+            Initialize();
+        }
+
+        static void Initialize()
+        {
+#if UNITY_EDITOR
+            InitializeInEditor();
+#else
+            InitializeInPlayer();
+#endif
+        }
+
+#if UNITY_EDITOR
+        internal static void InitializeInEditor()
+        {
+            var settings = ScriptableObject.CreateInstance<MLAgentsSettings>();
+            if (EditorBuildSettings.TryGetConfigObject(EditorBuildSettingsConfigKey,
+                out MLAgentsSettings settingsAsset))
+            {
+                if (settingsAsset != null)
+                {
+                    settings = settingsAsset;
+                }
+            }
+            Settings = settings;
+        }
+#else
+        internal static void InitializeInPlayer()
+        {
+            s_Settings = Resources.FindObjectsOfTypeAll<MLAgentsSettings>().FirstOrDefault() ?? ScriptableObject.CreateInstance<MLAgentsSettings>();
```

Comment:
```suggestion
            Settings = Resources.FindObjectsOfTypeAll<MLAgentsSettings>().FirstOrDefault() ?? ScriptableObject.CreateInstance<MLAgentsSettings>();
```
?

Reply:
good catch. The callback in Settings setter is for refreshing in the editor so should not really matter though.

Reply:
Can you explain in comments why we have both `Settings` and `s_Settings`?

Reply:
added comments on the getter/setter